lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	defer context.AfterFunc(s.runCtx, func() { _ = c.CloseWithError(0, "") })()
+
+	for {
+		// TODO(espadolini): stop accepting new streams once s.serveCtx is
+		// canceled, once quic-go gains the ability to change the amount of
+		// available streams during a connection (so we can set it to 0)
+		st, err := c.AcceptStream(context.Background())
+		if err != nil {
+			log.DebugContext(c.Context(), "Got an error accepting a stream.", "error", err)
+			return
+		}
+
+		s.wg.Add(1)
+		go s.handleStream(st, c, log)
+	}
+}
+
+func (s *QUICServer) handleStream(st quic.Stream, c quic.EarlyConnection, log *slog.Logger) {
+	defer s.wg.Done()
+
+	log = log.With("stream_id", st.StreamID())
+	defer log.DebugContext(c.Context(), "Done handling stream.")
+
+	defer st.CancelRead(0)
+	defer st.CancelWrite(0)
+
+	log.DebugContext(c.Context(), "Handling stream.")
+
+	sendErr := func(toSend error) {
+		st.CancelRead(0)
+		defer st.CancelWrite(0)
+		errBuf, err := proto.Marshal(&quicpeeringv1a.DialResponse{
+			Status: status.Convert(trail.ToGRPC(toSend)).Proto(),
+		})
+		if err != nil {
+			return
+		}
+		st.SetWriteDeadline(time.Now().Add(10 * time.Second))
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, uint32(len(errBuf)))); err != nil {
+			return
+		}
+		if _, err := st.Write(errBuf); err != nil {
+			return
+		}
+		if err := st.Close(); err != nil {
+			return
+		}
+	}
+
+	st.SetReadDeadline(time.Now().Add(10 * time.Second))
+	var reqLen uint32
+	if err := binary.Read(st, binary.LittleEndian, &reqLen); err != nil {
+		log.DebugContext(c.Context(), "Failed to read request size.", "error", err)
+		return
+	}
+	if reqLen >= quicMaxMessageSize {
+		log.WarnContext(c.Context(), "Received oversized request.", "request_len", reqLen)
+		return
+	}
+	reqBuf := make([]byte, reqLen)
+	if _, err := io.ReadFull(st, reqBuf); err != nil {
+		log.DebugContext(c.Context(), "Failed to read request.", "error", err)
+		return
+	}
+	st.SetReadDeadline(time.Time{})
+
+	req := new(quicpeeringv1a.DialRequest)
+	if err := proto.Unmarshal(reqBuf, req); err != nil {
+		log.WarnContext(c.Context(), "Failed to unmarshal request.", "error", err)
+		return
+	}
+
+	if requestTimestamp := req.GetTimestamp().AsTime(); time.Since(requestTimestamp).Abs() > quicTimestampGraceWindow {
+		log.WarnContext(c.Context(),
+			"Dial request has out of sync timestamp, 0-RTT performance will be impacted.",
+			"request_timestamp", requestTimestamp,
+		)
+		select {
+		case <-c.HandshakeComplete():
+		case <-c.Context().Done():
+			return
+		}
+	}
+
+	// a replayed request is always wrong even after a full handshake, the
+	// replay might've happened before the legitimate request
+	if !s.replayStore.add(req.GetNonce(), time.Now()) {
+		log.ErrorContext(c.Context(), "Request is reusing a nonce, rejecting.", "nonce", req.GetNonce())
+		sendErr(trace.BadParameter("reused or invalid nonce"))
+		return
+	}
+
+	_, clusterName, ok := strings.Cut(req.GetServerId(), ".")
```

Comment:
Please document the expected format of server_id in the proto.

Reply:
Logic copied wholesale from gRPC proxy peering.